signer/core/api.go

```diff
@@ -542,18 +575,153 @@ func (api *SignerAPI) Sign(ctx context.Context, addr common.MixedcaseAddress, da
+	if err != nil {
+		return nil, err
+	}
+	switch mediaType {
+	case ApplicationClique.Mime:
+		// Clique is the Ethereum PoA standard
+		header := &types.Header{}
+		if err := rlp.DecodeBytes(data, header); err != nil {
+			return nil, err
+		}
+		sighash, err := SignCliqueHeader(header)
+		if err != nil {
+			return nil, err
+		}
+		msg := fmt.Sprintf("Clique block %d [0x%x]", header.Number, header.Hash())
+		req = &SignDataRequest{Rawdata: data, Message: msg, Hash: sighash, ContentType: mediaType}
+	case ApplicationValidator.Mime:
+		// Data with an intended validator
+
+		sighash, msg := SignDataWithValidator(data)
+		req = &SignDataRequest{Rawdata: data, Message: msg, Hash: sighash, ContentType: mediaType}
+	case DataStructured.Mime:
+		// Typed data according to EIP712
+
+		sighash, msg := SignDataStructured(data)
+		req = &SignDataRequest{Rawdata: data, Message: msg, Hash: sighash, ContentType: mediaType}
+	case DataPlain.Mime:
+		// Sign calculates an Ethereum ECDSA signature for:
+		// keccack256("\x19${byte version}Ethereum Signed Message:\n" + len(message) + message))
+
+		// In the cases where it matters ensure that the charset is handled. The charset
+		// resides in the 'params' returned as the second returnvalue from mime.ParseMediaType
+		// charset, ok := params["charset"]
+		// As it is now, we accept any charset and just treat it as 'raw'.
+
+		sighash, msg := SignDataPlain(data)
+		req = &SignDataRequest{Rawdata: data, Message: msg, Hash: sighash, ContentType: mediaType}
+	default:
+		return nil, fmt.Errorf("content type '%s' not implemented for signing", contentType)
+	}
+	return req, nil
+
+}
+
+// SignCliqueHeader returns the hash which is used as input for the proof-of-authority
+// signing. It is the hash of the entire header apart from the 65 byte signature
+// contained at the end of the extra data.
+//
+// The method requires the extra data to be at least 65 bytes -- the original implementation
+// in clique.go panics if this is the case, thus it's been reimplemented here to avoid the panic
+// and simply return an error instead
+func SignCliqueHeader(header *types.Header) (hexutil.Bytes, error) {
+	hash := common.Hash{}
+	if len(header.Extra) < 65 {
+		return hash.Bytes(), fmt.Errorf("clique header extradata too short, %d < 65", len(header.Extra))
+	}
+	hasher := sha3.NewKeccak256()
+	rlp.Encode(hasher, []interface{}{
+		header.ParentHash,
+		header.UncleHash,
+		header.Coinbase,
+		header.Root,
+		header.TxHash,
+		header.ReceiptHash,
+		header.Bloom,
+		header.Difficulty,
+		header.Number,
+		header.GasLimit,
+		header.GasUsed,
+		header.Time,
+		header.Extra[:len(header.Extra)-65],
+		header.MixDigest,
+		header.Nonce,
+	})
+	hasher.Sum(hash[:0])
+	return hash.Bytes(), nil
+}
+
+// DataWithValidatorHash signs the given message which can be further recovered
+// with the given validator.
+func SignDataWithValidator(data []byte) ([]byte, string) {
+	msg := "TODO"
```

Comment:
I guess this is one of the hard parts -- figuring out what to show the user

Reply:
Well, I reckon EIP-712 is the one solving this issue. Shouldn't EIP-191 be an expansive standard, while 712 exclusive?

For this type, `0x00`, displaying the following two fields for the user:

- data <hex>
- validator <address>

Seems okay to me.